src/coreclr/nativeaot/Runtime/thread.cpp

```diff
@@ -1133,16 +1133,19 @@ FORCEINLINE bool Thread::InlineTryFastReversePInvoke(ReversePInvokeFrame * pFram
     // a do not trigger mode.  The exception to the rule allows us to have [UnmanagedCallersOnly] methods that are called via
     // the "restricted GC callouts" as well as from native, which is necessary because the methods are CCW vtable
     // methods on interfaces passed to native.
-    if (IsCurrentThreadInCooperativeMode())
+    // We will allow threads in DoNotTriggerGc mode to do reverse PInvoke regardless of their coop state.
+    if (IsDoNotTriggerGcSet())
     {
-        if (IsDoNotTriggerGcSet())
-        {
-            // RhpTrapThreads will always be set in this case, so we must skip that check.  We must be sure to
-            // zero-out our 'previous transition frame' state first, however.
-            pFrame->m_savedPInvokeTransitionFrame = NULL;
-            return true;
-        }
+        // We expect this scenario only when EE is stopped.
+        ASSERT(ThreadStore::IsTrapThreadsRequested());
```

Comment:
Is this correct? 

The stacktrace from https://github.com/dotnet/runtime/pull/85235#issuecomment-1522597237 suggests that the handle table scanning can be running concurrently in the background.

Reply:
That must be in the blocking phase that mops up what concurrent phase could have missed. 
Concurrent marking should not run callouts.

Reply:
You are right: https://github.com/dotnet/runtime/blob/add51b82baf268c3678bdd124480a46b89d72ce4/src/coreclr/gc/objecthandle.cpp#L1078

(I think it would be ok to do that, but it is not worth it to spend time chasing down the race conditions as the comment says.)

Reply:
Yes, there would be races with refcounts changing.

Also note that while callouts do not run concurrently with the program, they may run in parallel - on server GC. I hope the callouts are prepared for that. I do not know to the minor detail what they do. 
It would also be nice if they do not take a global lock, or something like that, since that can have serializing effect on parallel GC while in blocking phase. That is a separate concern from this PR though. As I understand these callouts are not expected to take much time anyways.